torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -1146,42 +1178,66 @@ class ShapePropagator {
     //   - has a bool keepdim argument
     static const register_formula_for dim_reduce_ops_with_integer_upcast{
         {
-            "aten::prod(Tensor self, int dim, bool keepdim) -> Tensor",
+            "aten::prod(Tensor self, int dim, bool keepdim, *, int? dtype) -> Tensor",
+        },
+        [](Node* node) -> type_vec_t {
+          auto maybe_keepdim = node->get<bool>(attr::keepdim);
+          at::optional<IValue> opt_dtype = node->get(attr::dtype);
+          return reduce_op_handler(
+              node, /*num_reduce_dim=*/*maybe_keepdim?0:1, /*integer_upcast=*/true, opt_dtype);
+        }};
+
+    // Requirements:
+    //   dims           : preserved
+    //   scalar type    : preserved unless specified.
+    //   device         : preserved
+    //   tensor inputs  : 1
+    //   tensor outputs : 1
+    // Additionally:
+    //   - First input should be the only tensor input
+    static const register_formula_for dim_reduce_ops_dtype{
+        {
+          "aten::cumprod(Tensor self, int dim, *, int? dtype) -> Tensor",
+          "aten::cumsum(Tensor self, int dim, *, int? dtype) -> Tensor",
+          "aten::log_softmax(Tensor self, int dim, *, int? dtype) -> Tensor"
         },
         [](Node* node) -> type_vec_t {
-          return multidim_reduce_with_postprocess(
-              node, /*num_reduce_dim=*/1, /*integer_upcast=*/true);
+          at::optional<IValue> opt_dtype = node->get(attr::dtype);
+          return reduce_op_handler(
+              node, /*num_reduce_dim=*/0, /*integer_upcast=*/true, opt_dtype);
         }};
 
     // Requirements:
     //   dims           : preserved if keepdim == false, dim->size() smaller
-    //   otherwise scalar type    : preserved device         : preserved tensor
-    //   inputs  : 1 tensor outputs : 1
+    //   otherwise
+    //   scalar type    : preserved
+    //   device         : preserved
+    //   tensor inputs  : 1
+    //   tensor outputs : 1
     // Additionally:
     //   - First input should be the only tensor input
     //   - has a bool keepdim argument
     static const register_formula_for multidim_reduce_ops{
         {
             "aten::logsumexp(Tensor self, int[] dim, bool keepdim) -> Tensor",
-            "aten::mean(Tensor self, int[] dim, bool keepdim) -> Tensor",
             "aten::norm(Tensor self, Scalar? p, int[] dim, bool keepdim) -> Tensor",
             "aten::std(Tensor self, int[] dim, bool unbiased, bool keepdim) -> Tensor",
             "aten::var(Tensor self, int[] dim, bool unbiased, bool keepdim) -> Tensor",
             "aten::max_values(Tensor self, int[] dim, bool keepdim) -> Tensor",
             "aten::min_values(Tensor self, int[] dim, bool keepdim) -> Tensor",
         },
         [](Node* node) -> type_vec_t {
-          if (auto dim = node->get<std::vector<int64_t>>(attr::dim)) {
-            return multidim_reduce_with_postprocess(
+          if (auto dims = node->namedInput(attr::dim)->node()->inputs().size()) {
```

Comment:
previous code used `auto dim = node->get<std::vector<int64_t>>(attr::dim)` and then `dim->size()`, which I believe was incorrect and rarely would have given a correct graph for these functions. In my debugging it always returned c10::nullopt. I believe this is because get() from ir.cpp calls `toIValue()` from constants.cpp, which returns `c10::nullopt` whenever `value->node()->kind() != prim::Constant`.